### ISSUE TITLE
Feature Request- Allow specification of a sorting function for entities

### ISSUE THREAD

**Issue by mxdmedia** [briebug/ngrx-auto-entity#58]
In `@ngrx/entity`, when building an adapter, one is allowed to provide a sort function to be applied whenever new entities are added to the store.  That sort function is used to sort the ids in the entity id list.  For example, if my model looks like this:

```typescript
export class MyModel {
  @Key id: string;
  name: string;
  delta: number;
}
```

delta is a value that determines the ordering of models when they're listed.  Thus, in `@ngrx/entity` I'd do the following when making my adaptor:

```typescript
import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
 
export interface MyModel {
  id: string;
  name: string;
  delta: number;
}

export function sortByDelta(a: MyModel, b: MyModel): number {
  return a.delta > b.delta ? 1 : -1;
}
 
export const adapter: EntityAdapter<MyModel> = createEntityAdapter<MyModel>({
  sortComparer: sortByDelta,
});
```

Then, since my ids are now ordered by the entity's delta, the pre-built "selectAll" selector (and others) will enforce the selector's output being ordered using the `sortByDelta` function.

Implementing this functionality in `auto-entity` would be fantastic.  One thought: would it be possible to use a decorator, rather than an explicit sorting function?  Something like:

```typescript
export class MyModel {
  @Key id: string;
  name: string;
  @OrderBy('asc') delta: number;
}
```

Where `'asc'` denotes sorting ascending, and `'desc'` would indicate a descending order. 
 One could also potentially do something similar to how multiple keys work.  Something like:

```typescript
export class User {
  @Key id: string;
  @OrderBy('asc', 2) firstName: string;
  @OrderBy('asc', 1) lastName: string;
}
```

where the second parameter in the `OrderBy` decorator is the sort order (e.g.- sort first by lastName, then by firstName).  I am not deeply familiar with writing decorators, so I don't know if something like this is even possible.

**Comment by jrista** [comment]
@mxdmedia I have researched this a bit more. Based on how the data is stored in state, unless the sort keys were also the primary keys of the entities, I don't think we will be able to actually store the entities in state in any particular sorted order. 

We can and will certainly be able to sort the entities according to the specified sort key(s) when selecting data from state. I just wanted to make sure you were aware that if you were looking for a potential performance improvement by STORING the entities in some particular sort order in state, that probably won't be possible. This is because we store the core state the same ways that @ngrx/entity does (in hopes of maintaining some compatibility there): 

```typescript
entities: {},
ids: []
```

So when we select these, we do so like this:

```typescript
selectAll: createSelector(entityState, state => state.entityName.ids.map(id => state.entityName.entities[id]))
```

I am still mulling over exactly how to support retrieving sorted entities. Sorting will add overhead, so I am wondering if the best approach, rather than augmenting existing selectors (which could impact existing implementations as well), would instead be to add new sorted selectors:

```typescript
selectAllSorted: createSelector(selectAll, entities => entities.sort(EntityModel.__ngrx_ae_defaultSortComparer))
```

Something along those lines. This would give you the flexibility to incur the hit to sort, or not, as appropriate for your use case.

**Comment by mxdmedia** [comment]
I'm not sure if this is helpful but, I believe the way `entity` does this is by re-sorting the `ids` array.  So you take a performance hit on any operations that add or modify entities, as the ids need to be resorted each time (a delete wouldn't require re-sorting, as the remaining entity ids should still be in the correct order).

Looking at the entity code, there are two possible entity adapters used- a sorted adapter (created if you supply a sort function), and an unsorted adapter (created if you don't supply a sort function).  For the sorted adapter, any time you mutate the entity state any new/modified entities are first sorted using the provided sort function.  Next, the existing entities (filtering out any entities with matching ids to the new/modified entities) are grabbed as an array (using the previously sorted order).  An ids array is then created, by merging the two entity arrays, using the sort comparison function at each step, and mapping the elements back their ids. 

Thoughts?

**Comment by jrista** [comment]
I will have to take a look and see. I guess the ids array could be sorted, I'll need to see how much that affects the complexity of the reducer overall. Aside from many different ways to load data (each with different semantics, such as pages vs. ranges, etc.; I am not sure sorting in the app would work at all with ranges...or, for that matter, paging), we would need to make sure that if any sort keys are modified on update/replace, and of course when new entities are created, we would need to maintain the sort order then as well.

That means the sorting hit would be incurred every time state changed for any reason... Thanks to memoization, sorting on selection only when sorting is desired might actually be more efficient. I'll keep you posted, as I think I'm going to do some testing to evaluate the performance of the various options.

**Comment by jrista** [comment]
@mxdmedia I am currently leaning fairly strongly in the direction of sorting in selectors. Due to the nature of loading with auto-entity, we can load data sets that may work well with sorting, and others that may not. 

So I think what I'll be doing is adding some selectors for retrieving sorted results, based on the entity decorator config. That way the sort cost is only incurred when you request it, and it will be memoized by the selector. If customization of the sorted set is needed, additional custom selectors could be written to compose auto-entity selectors.

**Comment by mxdmedia** [comment]
@jrista 

Sounds great.  To get an idea of how it would work..  you'd provide a sort function with a decorator, and that would automagically be imported/used in an auto-entity generated `selectSorted` selector? I'm imagining something like this:

`person.model.ts`
```typescript
const sortFn = (a, b) => { return a === b ? 0 : a > b ? 1 : -1 };

@EntitySorter('Person')
export const sortByName = (a: Person, b: Person) => {
  return sortFn(a.lastName,b.lastName) || sortFn(a.firstName,b.firstNname);
}

export class Person {
  @Key id: string;
  firstName: string;
  lastName: string;
}
```

**Comment by jrista** [comment]
Close, it will be decorator based, but it will be integrated into the `@Entity` decorator:

```typescript
// This comparison stuff could be general purpose, if written more generally:
export interface IHasFirstLast {
    firstName: string;
    lastName: string;
}

const compareFn = (a, b) => a === b ? 0 : a > b ? 1 : -1;

export const compareByLastFirst = (a: IHasFirstLast, b: IHasFirstLast) => 
    compareFn(a.lastName, b.lastName) || compareFn(a.firstName, b.firstName);

// Entity model plugs in a comparer with the @Entity decorator:
@Entity({
    pluralName: 'People',
    comparer: compareByLastFirst
})
export class Person {
    @Key id: string;
    firstName: string;
    lastName: string;
}
```

**Comment by jrista** [comment]
Resolved by #84
